File: packages/spectacular-tools/src/tree.rs
use crate::display::tool_arg_line;
use crate::output_preview::preview_text;
use crate::path::resolve_workspace_path;
use glob::Pattern;
use ignore::gitignore::{Gitignore, GitignoreBuilder};
use serde::Deserialize;
use serde_json::{json, Value};
use spectacular_agent::{Cancellation, Tool, ToolDisplay, ToolExecution, ToolManifest};
use std::path::{Path, PathBuf};

pub const TREE_TOOL_NAME: &str = "tree";

const HIDDEN_EXCEPTIONS: &[&str] = &[".agents"];

const TREE_TOOL_DESCRIPTION: &str = "Display directory structure as an ASCII tree. Directories are listed first, then files, both sorted alphabetically. Respects .gitignore and excludes hidden files except .agents.";

#[derive(Clone, Debug)]
pub struct TreeTool {
    workspace_root: PathBuf,
}

impl TreeTool {
    /// Creates a tree tool scoped to the provided workspace root.
    pub fn new(workspace_root: impl Into<PathBuf>) -> Self {
        Self {
            workspace_root: workspace_root.into(),
        }
    }
}

impl Tool for TreeTool {
    /// Returns the stable tool name used for registration and dispatch.
    fn name(&self) -> &str {
        TREE_TOOL_NAME
    }

    /// Builds the tree tool manifest and JSON parameter schema.
    fn manifest(&self) -> ToolManifest {
        ToolManifest::new(
            TREE_TOOL_NAME,
            TREE_TOOL_DESCRIPTION,
            json!({
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Root directory to display (default: current directory). Relative paths resolve against the workspace root; absolute paths and .. traversal are allowed intentionally."
                    },
                    "exclude": {
                        "type": "array",
                        "items": { "type": "string" },
                        "description": "Glob patterns to exclude from the tree"
                    }
                },
                "additionalProperties": false
            }),
        )
    }

    /// Formats tree arguments as the requested root path.
    fn format_input(&self, arguments: &Value) -> ToolDisplay {
        let path = arguments.get("path").and_then(Value::as_str).unwrap_or(".");
        path.to_owned()
    }

    /// Formats tree arguments as a styled renderer call line.
    fn format_call(&self, arguments: &Value) -> ToolDisplay {
        let path = arguments.get("path").and_then(Value::as_str).unwrap_or(".");
        tool_arg_line("List", path)
    }

    /// Formats tree output as a bounded text preview.
    fn format_output(&self, raw_output: &str, _parsed_output: Option<&Value>) -> ToolDisplay {
        preview_text(raw_output)
    }

    /// Executes tree rendering and returns the plain-text tree output.
    fn execute<'a>(&'a self, arguments: Value, _cancellation: Cancellation) -> ToolExecution<'a> {
        let workspace_root = self.workspace_root.clone();

        Box::pin(async move {
            let input = match serde_json::from_value::<TreeInput>(arguments) {
                Ok(input) => input,
                Err(error) => {
                    return Ok(format!("Error: invalid input JSON: {error}"));
                }
            };

            Ok(execute_tree(&workspace_root, input))
        })
    }
}

#[derive(Debug, Deserialize)]
struct TreeInput {
    #[serde(default)]
    path: Option<String>,
    #[serde(default)]
    exclude: Option<Vec<String>>,
}

/// Resolves tree input, validates the root, and renders the directory tree.
fn execute_tree(workspace_root: &Path, input: TreeInput) -> String {
    let raw_path = input.path.unwrap_or_default();
    let display_path = if raw_path.is_empty() { "." } else { &raw_path };
    let root = resolve_workspace_path(workspace_root, display_path);

    if !root.exists() {
        return format!("Error: path not found: {display_path}");
    }

    if !root.is_dir() {
        return format!("Error: path is not a directory: {display_path}");
    }

    let exclude_patterns = match parse_exclude_patterns(input.exclude.unwrap_or_default()) {
        Ok(patterns) => patterns,
        Err(error) => return error,
    };

    let root_name = root.file_name().unwrap_or_default().to_string_lossy();
    if exclude_patterns
        .iter()
        .any(|pattern| pattern.matches(&root_name))
    {
        return format!("Error: exclude pattern matches the root directory: {root_name}");
    }

    let canonical = root.canonicalize().unwrap_or_else(|_| root.clone());
    let root_display = canonical
        .to_string_lossy()
        .replace('\\', "/")
        .trim_start_matches("//?/")
        .to_string();

    let mut gitignores = Vec::new();
    let mut output = format!("{root_display}\n");
    build_tree(&root, "", &mut gitignores, &exclude_patterns, &mut output);
    output
}

/// Parses exclude glob patterns and returns a user-facing error for invalid patterns.
fn parse_exclude_patterns(patterns: Vec<String>) -> Result<Vec<Pattern>, String> {
    patterns
        .into_iter()
        .map(|pattern| {
            Pattern::new(&pattern)
                .map_err(|error| format!("Error: invalid exclude pattern '{pattern}': {error}"))
        })
        .collect()
}

/// Recursively appends sorted visible entries to the ASCII tree output.
fn build_tree(
    dir: &Path,
    prefix: &str,
    gitignores: &mut Vec<Gitignore>,
    exclude_patterns: &[Pattern],
    output: &mut String,
) {
    let added = maybe_push_gitignore(dir, gitignores);
    let entries = match filtered_sorted_entries(dir, gitignores, exclude_patterns) {
        Ok(entries) => entries,
        Err(_) => {
            if added {
                gitignores.pop();
            }
            return;
        }
    };

    for (index, (path, is_dir)) in entries.iter().enumerate() {
        let is_last = index == entries.len() - 1;
        let connector = if is_last { "`-- " } else { "|-- " };
        let name = path.file_name().unwrap_or_default().to_string_lossy();

        if *is_dir {
            output.push_str(&format!("{prefix}{connector}{name}/"));
            let mark = output.len();
            output.push('\n');

            let child_prefix = if is_last {
                format!("{prefix}    ")
            } else {
                format!("{prefix}|   ")
            };
            build_tree(path, &child_prefix, gitignores, exclude_patterns, output);

            if output.len() == mark + 1 {
                output.truncate(mark);
                output.push_str(" (empty)\n");
            }
            continue;
        }

        output.push_str(&format!("{prefix}{connector}{name}\n"));
    }

    if added {
        gitignores.pop();
    }
}

/// Reads visible directory entries, applies ignore/exclude rules, and sorts directories before files.
fn filtered_sorted_entries(
    dir: &Path,
    gitignores: &[Gitignore],
    exclude_patterns: &[Pattern],
) -> std::io::Result<Vec<(PathBuf, bool)>> {
    let mut dirs = Vec::new();
    let mut files = Vec::new();

    for entry in std::fs::read_dir(dir)? {
        let entry = match entry {
            Ok(entry) => entry,
            Err(_) => continue,
        };
        let path = entry.path();
        let name = path.file_name().unwrap_or_default().to_string_lossy();

        if name.starts_with('.') && !HIDDEN_EXCEPTIONS.contains(&name.as_ref()) {
            continue;
        }

        if is_symlink(&path) {
            continue;
        }

        let is_dir = path.is_dir();
        if is_gitignored(&path, is_dir, gitignores) || is_excluded(&path, exclude_patterns) {
            continue;
        }

        if is_dir {
            dirs.push(path);
        } else {
            files.push(path);
        }
    }

    dirs.sort_by_key(|path| case_insensitive_name(path));
    files.sort_by_key(|path| case_insensitive_name(path));

    let mut entries = Vec::with_capacity(dirs.len() + files.len());
    entries.extend(dirs.into_iter().map(|path| (path, true)));
    entries.extend(files.into_iter().map(|path| (path, false)));
    Ok(entries)
}

/// Reports whether a path is a symbolic link that should be skipped.
fn is_symlink(path: &Path) -> bool {
    std::fs::symlink_metadata(path)
        .map(|metadata| metadata.file_type().is_symlink())
        .unwrap_or(false)
}

/// Reports whether a path matches any user-supplied exclude pattern.
fn is_excluded(path: &Path, exclude_patterns: &[Pattern]) -> bool {
    if exclude_patterns.is_empty() {
        return false;
    }

    let name = path.file_name().unwrap_or_default().to_string_lossy();
    let path_str = path.to_string_lossy().replace('\\', "/");
    exclude_patterns
        .iter()
        .any(|pattern| pattern.matches(&name) || pattern.matches(&path_str))
}

/// Reports whether stacked gitignore matchers ignore the given path.
fn is_gitignored(path: &Path, is_dir: bool, gitignores: &[Gitignore]) -> bool {
    for gitignore in gitignores.iter().rev() {
        match gitignore.matched(path, is_dir) {
            ignore::Match::Ignore(_) => return true,
            ignore::Match::Whitelist(_) => return false,
            ignore::Match::None => continue,
        }
    }
    false
}

/// Loads a directory-local .gitignore onto the matcher stack when present.
fn maybe_push_gitignore(dir: &Path, gitignores: &mut Vec<Gitignore>) -> bool {
    let gitignore_path = dir.join(".gitignore");
    if !gitignore_path.is_file() {
        return false;
    }

    let mut builder = GitignoreBuilder::new(dir);
    builder.add(gitignore_path);
    match builder.build() {
        Ok(gitignore) => {
            gitignores.push(gitignore);
            true
        }
        Err(_) => false,
    }
}

/// Returns a lowercase file name used for stable case-insensitive sorting.
fn case_insensitive_name(path: &Path) -> String {
    path.file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_lowercase()
}

#[cfg(test)]
mod tests {
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/unit/tree.rs"));
}

File: tools/tree/src/lib/tree.ts
import { posix as path } from 'node:path';

import type { SandboxSession } from 'sandbox';
import { createTool as defineTool } from 'tools';
import { z } from 'zod';

const HIDDEN_EXCEPTIONS = new Set(['.agents']);

const description =
  'Display directory structure as an ASCII tree. Directories are listed first, then files, both sorted alphabetically. Respects .gitignore and excludes hidden files except .agents.';

export const schema = z
  .object({
    path: z.string().optional(),
    exclude: z.array(z.string()).optional(),
  })
  .strict();

type Options = {
  readonly workspaceRoot: string;
  readonly sandbox: SandboxSession;
};

type Entry = {
  readonly path: string;
  readonly isDirectory: boolean;
};

type IgnorePattern = {
  readonly base: string;
  readonly pattern: string;
  readonly negated: boolean;
};

type PathKind = 'directory' | 'file' | 'missing' | 'other';

/** Creates the provider-neutral ASCII tree tool. */
export const createTool = ({ workspaceRoot, sandbox }: Options) =>
  defineTool({
    name: 'tree',
    description,
    schema,
    execute: (input): Promise<string> => execute(workspaceRoot, sandbox, input),
  });

const execute = async (
  workspaceRoot: string,
  sandbox: SandboxSession,
  input: z.output<typeof schema>,
): Promise<string> => {
  const rawPath = input.path ?? '';
  const displayPath = rawPath === '' ? '.' : rawPath;
  const root = resolvePath(workspaceRoot, displayPath);

  if (typeof root === 'string') {
    return `Error: ${root}`;
  }

  const kind = await pathKind(sandbox, workspaceRoot, root.path);
  if (kind === 'missing') {
    return `Error: path not found: ${displayPath}`;
  }

  if (kind !== 'directory') {
    return `Error: path is not a directory: ${displayPath}`;
  }

  const excludes = input.exclude?.map(compileGlob) ?? [];
  const invalid = excludes.find(
    (value): value is string => typeof value === 'string',
  );
  if (invalid !== undefined) {
    return invalid;
  }

  const rootName = path.basename(root.path);
  if ((excludes as RegExp[]).some((pattern) => pattern.test(rootName))) {
    return `Error: exclude pattern matches the root directory: ${rootName}`;
  }

  const entries = await visibleEntries(
    sandbox,
    workspaceRoot,
    root.path,
    excludes as RegExp[],
  );
  const children = groupEntries(entries);
  const lines = [root.path];
  appendTree(root.path, '', children, lines);

  return `${lines.join('\n')}\n`;
};

const appendTree = (
  dir: string,
  prefix: string,
  children: ReadonlyMap<string, readonly Entry[]>,
  lines: string[],
): void => {
  const entries = children.get(dir) ?? [];

  for (const [index, entry] of entries.entries()) {
    const isLast = index === entries.length - 1;
    const connector = isLast ? '`-- ' : '|-- ';
    const name = path.basename(entry.path);

    if (!entry.isDirectory) {
      lines.push(`${prefix}${connector}${name}`);
      continue;
    }

    const mark = lines.length;
    lines.push(`${prefix}${connector}${name}/`);
    const childPrefix = isLast ? `${prefix}    ` : `${prefix}|   `;
    appendTree(entry.path, childPrefix, children, lines);

    if (lines.length === mark + 1) {
      lines[mark] = `${lines[mark]} (empty)`;
    }
  }
};

const visibleEntries = async (
  sandbox: SandboxSession,
  workspaceRoot: string,
  root: string,
  excludes: readonly RegExp[],
): Promise<readonly Entry[]> => {
  const [directories, files, ignores] = await Promise.all([
    listPaths(sandbox, workspaceRoot, root, 'd'),
    listPaths(sandbox, workspaceRoot, root, 'f'),
    readIgnores(sandbox, workspaceRoot, root),
  ]);
  const entries = [
    ...directories
      .filter((entry) => entry !== root)
      .map((entry) => ({ path: entry, isDirectory: true })),
    ...files.map((entry) => ({ path: entry, isDirectory: false })),
  ];

  return entries.filter(
    (entry) =>
      !isHidden(root, entry.path) &&
      !isIgnoredWithAncestors(root, entry.path, entry.isDirectory, ignores) &&
      !isExcluded(entry.path, excludes),
  );
};

const groupEntries = (
  entries: readonly Entry[],
): ReadonlyMap<string, readonly Entry[]> => {
  const groups = new Map<string, Entry[]>();

  for (const entry of entries) {
    const parent = path.dirname(entry.path);
    groups.set(parent, [...(groups.get(parent) ?? []), entry]);
  }

  return new Map(
    [...groups.entries()].map(([parent, values]) => [parent, sorted(values)]),
  );
};

const sorted = (entries: readonly Entry[]): readonly Entry[] =>
  [...entries].sort((left, right) => {
    if (left.isDirectory !== right.isDirectory) {
      return left.isDirectory ? -1 : 1;
    }

    return path
      .basename(left.path)
      .localeCompare(path.basename(right.path), undefined, {
        sensitivity: 'accent',
      });
  });

const listPaths = async (
  sandbox: SandboxSession,
  workspaceRoot: string,
  root: string,
  type: 'd' | 'f',
): Promise<readonly string[]> => {
  const result = await sandbox.exec({
    cwd: workspaceRoot,
    cmd: [
      'find',
      root,
      '(',
      '-name',
      '.git',
      ')',
      '-prune',
      '-o',
      '-type',
      type,
      '-print',
    ],
  });

  if (result.exitCode !== 0) {
    return [];
  }

  return lines(result.stdout).map(normalizePath).sort();
};

const readIgnores = async (
  sandbox: SandboxSession,
  workspaceRoot: string,
  root: string,
): Promise<readonly IgnorePattern[]> => {
  const result = await sandbox.exec({
    cwd: workspaceRoot,
    cmd: [
      'find',
      root,
      '(',
      '-name',
      '.git',
      ')',
      '-prune',
      '-o',
      '-type',
      'f',
      '-name',
      '.gitignore',
      '-print',
    ],
  });

  if (result.exitCode !== 0) {
    return [];
  }

  const files = lines(result.stdout).map(normalizePath).sort();
  const groups = await Promise.all(
    files.map(async (file) =>
      parseIgnores(
        path.dirname(file),
        await sandbox.readFile(file).catch(() => ''),
      ),
    ),
  );

  return groups.flat();
};

const parseIgnores = (base: string, text: string): readonly IgnorePattern[] =>
  text
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line !== '' && !line.startsWith('#'))
    .map((line) => ({
      base,
      pattern: line.startsWith('!') ? line.slice(1) : line,
      negated: line.startsWith('!'),
    }))
    .filter((ignore) => ignore.pattern !== '');

const isHidden = (root: string, fullPath: string): boolean =>
  path
    .relative(root, fullPath)
    .split('/')
    .some((part) => part.startsWith('.') && !HIDDEN_EXCEPTIONS.has(part));

const isIgnoredWithAncestors = (
  root: string,
  fullPath: string,
  isDirectory: boolean,
  ignores: readonly IgnorePattern[],
): boolean => {
  const ignoredAncestor = ancestors(root, fullPath).some((ancestor) =>
    isIgnored(ancestor, true, ignores),
  );

  return ignoredAncestor || isIgnored(fullPath, isDirectory, ignores);
};

const ancestors = (root: string, fullPath: string): readonly string[] => {
  const values: string[] = [];
  let current = path.dirname(fullPath);

  while (contains(root, current) && current !== root) {
    values.unshift(current);
    current = path.dirname(current);
  }

  return values;
};

const isIgnored = (
  fullPath: string,
  isDirectory: boolean,
  ignores: readonly IgnorePattern[],
): boolean => {
  let ignored = false;
  for (const ignore of ignores) {
    const pattern = ignore.pattern.endsWith('/')
      ? ignore.pattern.slice(0, -1)
      : ignore.pattern;
    if (ignore.pattern.endsWith('/') && !isDirectory) {
      continue;
    }

    if (!contains(ignore.base, fullPath)) {
      continue;
    }

    const glob = compileGlob(pattern);
    const relative = path.relative(ignore.base, fullPath);
    const matched =
      glob instanceof RegExp
        ? glob.test(relative) || glob.test(path.basename(fullPath))
        : relative === pattern || path.basename(fullPath) === pattern;
    if (matched) {
      ignored = !ignore.negated;
    }
  }

  return ignored;
};

const isExcluded = (fullPath: string, excludes: readonly RegExp[]): boolean => {
  const name = path.basename(fullPath);
  return excludes.some(
    (pattern) => pattern.test(name) || pattern.test(fullPath),
  );
};

const compileGlob = (pattern: string): RegExp | string => {
  if (pattern.includes('[') || pattern.includes(']')) {
    const reason =
      pattern.includes('[') && !pattern.includes(']')
        ? 'unclosed character class'
        : 'unsupported character classes';
    return `Error: invalid exclude pattern '${pattern}': ${reason}`;
  }

  const source = pattern
    .replace(/[\\^$+?.()|{}]/g, '\\$&')
    .replaceAll('**', '\0')
    .replaceAll('*', '[^/]*')
    .replaceAll('?', '[^/]')
    .replaceAll('\0', '.*');

  return new RegExp(`^${source}$`);
};

const resolvePath = (
  workspaceRoot: string,
  value: string,
): { readonly path: string } | string => {
  const root = normalizePath(workspaceRoot);
  const resolved = normalizePath(
    path.isAbsolute(value) ? value : path.join(root, value),
  );

  if (!contains(root, resolved)) {
    return `Path escapes workspace: ${value}`;
  }

  return { path: resolved };
};

const pathKind = async (
  sandbox: SandboxSession,
  workspaceRoot: string,
  value: string,
): Promise<PathKind> => {
  const result = await sandbox.exec({
    cwd: workspaceRoot,
    cmd: [
      'sh',
      '-c',
      'if [ -d "$1" ]; then printf directory; elif [ -f "$1" ]; then printf file; elif [ -e "$1" ]; then printf other; else printf missing; fi',
      'sh',
      value,
    ],
  });

  if (result.exitCode !== 0) {
    return 'missing';
  }

  return parsePathKind(result.stdout);
};

const parsePathKind = (value: string): PathKind => {
  const normalized = value.trim();

  if (
    normalized === 'directory' ||
    normalized === 'file' ||
    normalized === 'missing' ||
    normalized === 'other'
  ) {
    return normalized;
  }

  return 'missing';
};

const contains = (root: string, child: string): boolean =>
  child === root || child.startsWith(`${root}/`);

const lines = (value: string): readonly string[] =>
  value.split(/\r?\n/).filter((line) => line !== '');

const normalizePath = (value: string): string => {
  const resolved = path.normalize(path.isAbsolute(value) ? value : `/${value}`);
  return resolved === '/' ? resolved : resolved.replace(/\/+$/, '');
};
